Activate ELMO functionality in Box Model

bnmurphy opened the pull request:
Setenvvar and setenvvarc are updated to successfully set environment variables. ELMO can now create output file names and proceed with output. These data are written to CSV files which upon first inspection, look correctly articulated.

Minor changes were also made to the destination of some warnings -- from the screen to the log file.

Minor changes were made to number formats in ELMO to reduce compiler warning messages.

Compile directives were added so that data extraction calls from ASX_DATA_MOD are skipped for variables that are not present or expected (e.g. WSPD10).

bnmurphy committed:
Enable ELMO and move some log output to log file
 On branch m3box_elmo
	modified:   CCTM/src/box_io/m3utilio.f
	modified:   CCTM/src/box_io/mio_setenvvarc.c
	modified:   CCTM/src/box_io/xtract_metgeodata.f
	modified:   CCTM/src/driver/ELMO_INIT_DEFN.F
	modified:   CCTM/src/vdiff/acm2/ASX_DATA_MOD.F

bhutzell commented on CCTM/src/vdiff/acm2/ASX_DATA_MOD.F:
```diff
@@ -957,15 +958,16 @@ Subroutine GET_MET ( JDATE, JTIME, TSTEP )
 
          call interpolate_var ('PRSFC', jdate, jtime, Met_Data%PRSFC)
-
+#ifndef m3box
```
I recommend removing the added `#ifndef m3box` directives. For one reason, the log messages regarding missing input variables allow future changes on what input variables can be to added. Another reason is that `interpolate_var` sets the missing variable to 1.0E-30. The value can prevent floating point errors such as on line 1132 where `MET_Data%USTAR` divides `-TMPFX`.

bnmurphy replied:
Ah, that's a great point. I assumed we would remove them in the future anyway, once dry deposition is able to be added back in. But yes, I'll remove them now.

bhutzell commented on CCTM/src/box_io/xtract_metgeodata.f:
```diff
@@ -168,14 +169,14 @@ LOGICAL FUNCTION XTRACT_METGEODATA( VNAME,VNAME_VALUE )
             IF (  VNAME .EQ. 'ZH'  ) THEN
                   VNAME_VALUE  = BXM_ZH
-                  print*,'XTRACT_METGEODATA: BXM_ZH,ZH = ',BXM_ZH,VNAME_VALUE
+                  WRITE( LOGDEV, '(A,F,2X,F)'),'XTRACT_METGEODATA: BXM_ZH,ZH = ',BXM_ZH,VNAME_VALUE
```
Perhaps, the lines should be deleted because they supported initial development and don't have a current usefulness. Also, the format `F` descriptor causes crashes when using the gcc compiler.

bnmurphy replied:
Done.

bnmurphy committed:
Removed m3box ifdefs in ASX_DATA_MOD and shifted log output for interpolate_Var to logdev
 On branch m3box_elmo
	modified:   CCTM/src/box_io/xtract_metgeodata.f
	modified:   CCTM/src/cio/centralized_io_module.F
	modified:   CCTM/src/vdiff/acm2/ASX_DATA_MOD.F

bnmurphy committed:
Merge branch 'boxmodel' into m3box_elmo

bnmurphy commented:
K, I think this all works correctly together now. I was able to reroute the interpolate_var log output to the log file instead of the screen. Is that okay?

Thanks for the tip on the ASX_DATA_MOD issue. Now I can run in DEBUG mode just fine. It takes ~2.9 seconds. What does it do on Atmos?

bhutzell commented:
> K, I think this all works correctly together now. I was able to reroute the interpolate_var log output to the log file instead of the screen. Is that okay?
> 
> Thanks for the tip on the ASX_DATA_MOD issue. Now I can run in DEBUG mode just fine. It takes ~2.9 seconds. What does it do on Atmos?

About the same using either ifort or ifx compiler, the NO DEBUG mode has less than a second runtime. Please reply on what STD compile options are you using the ifx compiler.

bnmurphy commented:
Here are my options for standard and debug:
FSTD = -O3 -fno-alias -fp-model source -ftz -align all -unroll
DBG  = -O0 -g -check bounds -fpe0 -fno-alias -ftrapuv -traceback

bhutzell commented:
On the atmos platform, the pull request's updates only ran successfully if only the changed files were checkout from **bnmurphy:m3box_elmo**.  [The commit](https://github.com/bhutzell/CMAQ/commit/0c16cc2a96a86618200d900f0c7aea47be3d8262) records the action.

The problem was that the box model failed reading the environment variable setting the CTM_ IPR_1 file. If PROCAN was turned off,  the box model crashed getting the environment variable for a following output file. Perhaps, @bnmurphy can resolve the problem or determine whether the ad hoc fix is acceptable but at least the box model now incorporates the pull request's changes.

bnmurphy commented:
Thanks @bhutzell. I'm not seeing the issue I think you are describing. When I run the box model with PROCAN turned on, the expected PA output files are produced for IRR and IPR. When I turn it off, they are not produced and the model still runs successfully and produces all expected output.

bhutzell commented:
I figured that atmos was causing the runtime crashes but cherry picked your updates because I lost hope on solving the problem **until today**. 

The problem came from the ENSTR subroutine in fake_ioapi.F. The GET_ENVIRONMENT_VARIABLE intrinsic was giving bad STATUS values (_I never have encountered such an instance_). The fix replaces GET_ENVIRONMENT_VARIABLE with getenv. The change was tested from intel, pgi and gcc compilers with and without debug options.

bhutzell commented:
@bnmurphy Thanks for the ELMO and other updates. Please inform me if the merge causes problems.

bnmurphy commented:
Everything looks great to me!
